modyn/selector/internal/selector_strategies/downsampling_strategies/downsampling_scheduler.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional, Tuple
+
+from modyn.selector.internal.selector_strategies.downsampling_strategies import AbstractDownsamplingStrategy
+from modyn.selector.internal.selector_strategies.downsampling_strategies.utils import instantiate_downsampler
+
+
+class DownsamplingScheduler:
+    def __init__(
+        self, downsampling_configs: list[dict], downsampling_thresholds: list[int], maximum_keys_in_memory: int
+    ):
+        self.dowsampling_configs: list[dict] = downsampling_configs
+        self.dowsampling_thresholds: list[int] = downsampling_thresholds
+        self.downsampler_index = 0
+        # transition from downsampler[i] to downsampler[i+1] happens before the thresholds[i]-th trigger
+
+        self._validate_threshold()
+
+        self.maximum_keys_in_memory = maximum_keys_in_memory
+
+        self.current_downsampler, self.next_threshold = self._get_new_downsampler_and_threshold()
+
+    def _validate_threshold(self) -> None:
+        if not len(self.dowsampling_configs) == len(self.dowsampling_thresholds) + 1:
+            raise ValueError("You must specify a threshold for each transition")
+        if not sorted(self.dowsampling_thresholds) == self.dowsampling_thresholds:
+            raise ValueError("Thresholds must be monotonically increasing")
+        if not len(self.dowsampling_thresholds) == len(set(self.dowsampling_thresholds)):
+            raise ValueError("Thresholds must be unique")
+
+    def _get_new_downsampler_and_threshold(self) -> Tuple[AbstractDownsamplingStrategy, Optional[int]]:
+        next_downsampler = instantiate_downsampler(
+            self.dowsampling_configs[self.downsampler_index], self.maximum_keys_in_memory
+        )
+        # after instantiating the last downsampler, we simply set to none the next threshold
+        next_threshold = (
+            self.dowsampling_thresholds[self.downsampler_index]
+            if self.downsampler_index < len(self.dowsampling_thresholds)
+            else None
+        )
+        return next_downsampler, next_threshold
+
+    def _update_downsampler_if_needed(self, next_trigger_id: int) -> None:
+        if self.next_threshold is not None and next_trigger_id >= self.next_threshold:
+            self.downsampler_index += 1
+            self.current_downsampler, self.next_threshold = self._get_new_downsampler_and_threshold()
+
+    def get_requires_remote_computation(self, next_trigger_id: int) -> bool:
+        self._update_downsampler_if_needed(next_trigger_id)
```

Comment:
This should not be here. A getter like that should not have a side effect of updating something because this implies we rely on implicit behavior that is not obvious to the caller. I am not sure what the reason for this is, but we cannot have an update in the getter. Rather, we need to do this somewhere explicitly

Reply:
OK I created a function (`inform_next_trigger`) and removed the call from the getter. Also, as it no longer receives any parameter I removed all the `get_XY` and transformed them into python-style using the @property decorator. So now every property looks like this 
```
@property
def downsampling_params(self) -> dict:
        return self.current_downsampler.downsampling_params
```

Reply:
Oh that sounds cool. I will check it out when I do the final review round